duckduckgo/Android
# Feature/ask user for rating

Task/Issue URL: https://app.asana.com/0/715106103902962/747319243916954
Tech Design URL: 
CC: 

**Description**:
Asks user if they are enjoying the app after some amount of usage.

If the user is enjoying the app, we ask for a rating. If they are not enjoying it, we ask for feedback to make it better.

**Conditions**

_Initial Request_
We only ask if the following criteria are all met
1. They have the `Play Store` app installed, and they installed the `DuckDuckGo` app through the `Play Store`. *
1. They must have performed a minimum number of searches (5)
1. They must have used the app on at least 3 days (the first time the app is run counts as day 1)

_Follow up request_
Additionally, we might ask them again after another few days of usage (a minimum of 4 days later). We will only ask again if the user declined to give rating/feedback the first time around.

\* To ease testing, in `DEBUG` builds we don't enforce that the app must have been sourced from the `Play Store`.

**Steps to test this PR**:
There's a lot of combinations to test in this PR. Grab a ☕️; it's not going to be a quick test 😩 

There are two dialogs - you always see the same first dialog but which dialog you see next depends on how you answer the first dialog.

Additionally, you might also see a second dialog after 4 days of usage since the first dialog was dismissed (but only if you decline to give rating / feedback).

The testing will also involve changing the device DATE to simulate multiple days' usage.

ℹ️ If you want to reduce the steps required, you can hack the variables in `AppEnjoymentPromptCriteria.kt` to reduce the number of days' usage for example, after you've tested the flow is correct for some of the combos.

1. Install the app, and perform 5 searches
1. Background the app, and restore it; verify you are NOT prompted yet
1. Set the date forward one day
1. Return to the app; verify you are NOT prompted yet
1. Set the date forward another day
1. Return to the app; verify you ARE presented with a dialog (you are on day 3 of usage at this point)
1. Choose that you are not happy
1. Verify you see the prompt asking for feedback
1. Choose to dismiss
1. Set the date forward another day, return to the app (repeat until you reach 4 days after the first prompt); verify you don't see the prompt until 4 days later
1. Return to the app; verify you DO now see the second prompt
1. Choose NOT HAPPY, and choose to give feedback; verify the feedback prompt appears

1. Test the above, but this time at step 7. indicate you ARE happy
1. Follow the rest, but this time on the second prompt choose to give a rating; verify the Play Store listing appears

1. Follow the steps from test 1, but this time give feedback straightway; verify that the second prompt never appears

1. Follow the steps from test 1, but this time give a rating straightaway; verify that the second prompt never appears

1. For all options for the second prompt, verify you never see the prompt again 
    1. set date forward / return to app (repeat a few days)
 

![screenshot_1549035043](https://user-images.githubusercontent.com/1336281/52132118-61a7ff00-2636-11e9-9ece-979614a3d487.png)

![screenshot_1549035067](https://user-images.githubusercontent.com/1336281/52132146-71274800-2636-11e9-8776-5a750cd5a4dc.png)

![screenshot_1549035094](https://user-images.githubusercontent.com/1336281/52132167-7b494680-2636-11e9-8dc3-c5fa5d808c96.png)


**Testing migration**
1. Install from `develop` and run the app; add a bookmark
1. Update to this app version and ensure the bookmark is still there


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
